### PR TITLE
Linting: Update linter to enforce ES 2018 features

### LINKED ISSUE #25185: Update linter to enforce ES features
### Description

We'll want to update the linter (#23763) to ensure that ES2018 syntax is used consistently in both source and addons (see #25172). Currently it accepts a browserslist rather than an ES version to ensure that experimental browser APIs are used responsibly. This doesn't replace discussion around the use of those features, though.

### Solution

I'd like to replace or add on (preference for the first) a linter plugin that checks against ES features only. This is a common enough thing for me to be comfortable writing one on top of a parser if we can't find something well supported. As this widely affects quality control, this should have great language testing coverage before running wild here.

### Alternatives

Transpilation was mentioned in #24237, but discarded in favor of lighter to no tooling. I'll want to echo that in choice of linter tooling around this issue.

### Additional context

I haven't considered updating browerslist as per https://github.com/evanw/esbuild/issues/121#issuecomment-646956379 since that would give false positives for browser APIs. As this is a moving target, I don't think it's appropriate to update mdcs.

### PR DESCRIPTION
**Pull request by epreston**
Fixed #25185.

**Description**

Update linter to enforce ES 2018 features.

This has four changes:
- add setting to configure linter to enforce 2018 features
- separate config for linting into config file.
- update the 3 header imports that were using 2020 features
- to maintain the same public interface, add export objects MathUtils, AnimationUtils, and DataUtils

The setting that does the work of enforcing 2018 features is:

```json
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module"
  },
```

The 'module' setting has been provided by the custom style plugin, it is duplicated here for clarity.  This also ensures that the style plugin is only providing styles rules.

This also overrides the environment setting from the custom plugin to ensure only 2018 globals are defined.  Browser and node settings are duplicated for clarity.  This also ensures that the style plugin is only providing styles rules.

```json
  "env": {
    "browser": true,
    "node": true,
    "es2018": true
  },
```

The lines in 'Three.js' using 2020 syntax were:

```js
export * as AnimationUtils from './animation/AnimationUtils.js';
export * as MathUtils from './math/MathUtils.js';
export * as DataUtils from './extras/DataUtils.js';
```

With an export object added, these become:

```js
export { AnimationUtils } from './animation/AnimationUtils.js';
export { MathUtils } from './math/MathUtils.js';
export { DataUtils } from './extras/DataUtils.js';
```

If the project wants to remove a dependency, we can copy the 15 lines of style settings from the plugin into the lint config file.

### CODE REVIEW

**Comment by epreston** [comment]
Looks like MathUtils, AnimationUtils, and DataUtils need to be updated to export their namespaces.

**Comment by epreston** [comment]
If the project wants to remove a dependency, copy the 15 lines of style settings into the lint config file.

File: src/math/MathUtils.js
Comment (marcofugaro) [inline src/math/MathUtils.js:319]: Hey, exporting as plain object breaks tree-shaking, we can't have that.

Having `export * as` is fine as it gets transpiled by rollup in the build step.

Comment (epreston) [reply src/math/MathUtils.js:319]: Let's look into that.  There are a number of objects of this style.  This is what it gets trans-piled to.  I have a tree-shake confirmation project published.

This is being done in 2 parts.  https://github.com/mrdoob/three.js/issues/25185

Comment (epreston) [reply src/math/MathUtils.js:319]: Update: https://github.com/epreston/three-treeshake-coverage 

Please pretend you don't see the other bits.  Need an hour to cleanup my changes and commit them.

Comment (epreston) [reply src/math/MathUtils.js:319]: Long story short:

> Having `export * as` is fine as it gets transpiled by rollup in the build step.

Yes, it becomes something, which is very similar to this, before it gets distributed and the final bundler (client application) does the shake.  If we want to improve the shake, we have API changes.  If we want to retain the API, distribute nearly the same bundle, and enforce 2018, we change the code.

Currently the tools are configured for 2020.  Setting the tools to 2018 and enforcing it is the first step so we actually improve the tree-shake instead of masking it.

Comment (epreston) [reply src/math/MathUtils.js:319]: As a confirmation take a look at what we ship now using the 2020 syntax.  I'm going to use the middle size one to reduce the reading.

The old 2020 line:

```js
export * as AnimationUtils from './animation/AnimationUtils.js';
```

Is currently being translated to:

```js
var AnimationUtils = /*#__PURE__*/Object.freeze({
	__proto__: null,
	arraySlice: arraySlice,
	convertArray: convertArray,
	flattenJSON: flattenJSON,
	getKeyframeOrder: getKeyframeOrder,
	isTypedArray: isTypedArray,
	makeClipAdditive: makeClipAdditive,
	sortedArray: sortedArray,
	subclip: subclip
});
```

This has been updated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

It gets translated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

Keeping it clean, the final bundler works its magics.  This is what I hope to demonstrate in the confirmation project.

Comment (epreston) [reply src/math/MathUtils.js:319]: I'm writing this all out because you are right, except if we use this particular 2020 shorthand syntax, which doesn't help the cause.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: > If we want to improve the shake, we have API changes.

Actually you're correct, the object in the final bundle doesn't get tree-shaken because it's an object, and budnlers can't tree-shake that. The only way to make it tree-shakable would be to do `export * from './animation/AnimationUtils.js';`, but I'm not sure maintainers want to go with this with this breaking change.

> except if we use this particular 2020 shorthand syntax, which doesn't help the cause.
There is no problem with this syntax, as it doesn't end up in production. What #25185 is all about is Optional chaining and Nullish coalescing (and of course class fields from ES2022).

I guess we can enforce ES2020 and disable just Optional chaining and Nullish coalescing.

Comment (mrdoob) [reply src/math/MathUtils.js:319]: Unrelated but...

Could be nice if the CI could check if a PR makes the build bigger/smaller and also if the tree-shaked lib is bigger/smaller.

Comment (epreston) [reply src/math/MathUtils.js:319]: @mrdoob its on the way.  Also circular reference check.

Comment (epreston) [reply src/math/MathUtils.js:319]: > I guess we can enforce ES2020 and disable just Optional chaining and Nullish coalescing.

At the moment, that is not possible.  With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: Thanks @epreston! That would be cool. You can get the tree-shaken lib size from `test/treeshake/index.bundle.min.js`.

Comment (LeviPesin) [reply src/math/MathUtils.js:319]: Also maybe it can report test coverage? I think many large projects use bots that report both build size and test coverage.

Comment (epreston) [reply src/math/MathUtils.js:319]: I really believe in 50% best case reduction if we make some really careful and clever moves.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: > At the moment, that is not possible. With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.

Why is it not possible?

Comment (epreston) [reply src/math/MathUtils.js:319]: > disable just Optional chaining

I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust.  Seems like half of most classes are just checking for null / undefined.



> Nullish coalescing

Yeah, that's hard to read, causes bugs, and I can't spell it correctly without looking it up.

Comment (epreston) [reply src/math/MathUtils.js:319]: Quick heads up, "circular reference checks" are not going to win me any friends.  Nodes has 22.  Analysis posted here: #25240

Comment (LeviPesin) [reply src/math/MathUtils.js:319]: > I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust. Seems like half of most classes are just checking for null / undefined.

We have already disabled it and all of its usages.

> Nodes has 22. Analysis posted here: https://github.com/mrdoob/three.js/issues/25240

The solution is also posted there -- I will file a PR soon.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: > I don't think anyone wants to disable "Optional Chaining" that would drop a massive amount of code from the library and make things very robust. Seems like half of most classes are just checking for null / undefined.

By [Optional chaning](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) I mean `?.` and currently it's only used on `GLTFExporter.js`.
Nullish coalescing is not used anywhere in three.js.

Comment (epreston) [reply src/math/MathUtils.js:319]: > > At the moment, that is not possible. With the agreed 2018 direction, the zero impact to API and treeshake, this is the best we have.
> 
> Why is it not possible?

Not exposed as a rule, looks like it requires a custom eslint parser to implement.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: Ah yeah, well we could use the rules from this plugin: https://github.com/keithamus/eslint-plugin-escompat

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: I see this as the best solution, because as mentioned here https://github.com/mrdoob/three.js/issues/25185#issuecomment-1422059604, the import/exports are not included in the enforcing.

Comment (epreston) [reply src/math/MathUtils.js:319]: That looks promising for the next step.  Thank you. 

The plan is to relax back to 2020 parsing with a set of rules to strip it back that we can turn on and off.

Can you repost that on https://github.com/mrdoob/three.js/issues/25185 so we don't lose it ?

Comment (epreston) [reply src/math/MathUtils.js:319]: > I see this as the best solution, because as mentioned here [#25185 (comment)](https://github.com/mrdoob/three.js/issues/25185#issuecomment-1422059604), the import/exports are not included in the enforcing.

Which solution ?  There's a few more bits to consider to enforce 2018, without setting the tools to enforce, but add extra config to "sort of" enforce 2018.

Comment (marcofugaro) [reply src/math/MathUtils.js:319]: ES2020 + `no-nullish-coalescing` and `no-optional-chaining` would be the best solution.

Comment (epreston) [reply src/math/MathUtils.js:319]: The project you linked is for things that can't be polyfilled and trans-piled.  Keep in mind, Babel isn't in the toolchain.  So we don't have full polyfill and trans-pile.  Your idea has changed my mind about how much effort the next step will take.   I think it is awesome and we actually completely agree if you read what I wrote yesterday.

People keep confusing what is in the bundle vs what code is maintained.  It really is productive to have a separate discussion on the code we read and share.  Then, have another conversation about what is included in the bundle how that is presented.

At the moment, we don't bundle to a language standard but a browser support standard.  The language features used look like a rough mixture of es6 which changes every quarter.  Talking about the bundle in terms of language features makes little sense when its defined in terms of browser support.

The issue I see coming up will be about perceptions, assumptions, and confusion.  We have a standard, but we deviate from the standard, but something will fix it, then finally realising that is not what shipped.

I want to make people happy so they can produce the exact same output and remove 20 lines from the project.  If that needs to be a birds nest of tooling config, so be it.  We can make it so people can use a few 2020 features in a context of 2018.  Not keen on doing the work then having to remediate what has crept in.  Its easier to test and rationalise about with a stable base going forward.

Comment (epreston) [reply src/math/MathUtils.js:319]: > ES2020 + `no-nullish-coalescing` and `no-optional-chaining` would be the best solution.

That is not a complete solution or a very wise one.

Comment (epreston) [reply src/math/MathUtils.js:319]: I'm sorry.  That was too short of a response.  

My goal was to patiently explain why the assumption on tree-shaking was wrong.  When this came up I felt overwhelmed about going into what's missing.  I had the same ideas and learned the hard way.

I'll make it up to you.  This is not who I want to be.

**Comment by epreston** [comment]
Anyone worried about tree-shaking:

As a confirmation take a look at what we ship now using the 2020 syntax.  I'm going to use the middle size one to reduce the reading.

The old 2020 line:

```js
export * as AnimationUtils from './animation/AnimationUtils.js';
```

Is currently being translated to:

```js
var AnimationUtils = /*#__PURE__*/Object.freeze({
	__proto__: null,
	arraySlice: arraySlice,
	convertArray: convertArray,
	flattenJSON: flattenJSON,
	getKeyframeOrder: getKeyframeOrder,
	isTypedArray: isTypedArray,
	makeClipAdditive: makeClipAdditive,
	sortedArray: sortedArray,
	subclip: subclip
});
```

This has been updated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

It gets translated to:

```js
const AnimationUtils = {
	arraySlice: arraySlice,
	convertArray: convertArray,
	isTypedArray: isTypedArray,
	getKeyframeOrder: getKeyframeOrder,
	sortedArray: sortedArray,
	flattenJSON: flattenJSON,
	subclip: subclip,
	makeClipAdditive: makeClipAdditive
};
```

Keeping it clean, the final bundler (client application) works its magics.  Its possible we get slightly better tree-shake using the 2018 syntax.  Keep in mind this is an isolated context, relating to a library (not client application), and a very specific import command.

In the old code, final bundler can only remove the whole object if its entirely unused, which will be created in any case (new or old code).  Tools can remove constants if they are entirely unused, so we have even ground.

**Comment by epreston** [comment]
For evaluation, a test rig of the different builds and tree shake effects can be found here:

https://github.com/epreston/three-treeshake-coverage

Comment and uncomment the different imports, build and inspect the results in the dist folder.

In all cases, `AnimationUtils` is correctly tree-shaken.

**Comment by donmccurdy** [comment]
Is it possible to configure the linter to enforce ES2018 features, but to ignore ES2020 import/export syntax? I don't think there is any problem with shipping those ES2020 exports, and might allow us to make a narrower change here.

If that's not possible, that's OK, but I didn't see that mentioned in the thread so far.

**Comment by epreston** [comment]
Anything's possible with a configuration and tooling.  We have great ideas for partial solutions going 2020 with restrictions.  I don't see a built in way to approach the problem from that direction (2018 with 2020 imports).

In all cases, we never ship these exports.  The code this replaces is a developer convenience to shave 20 lines off the code we maintain.  The code it generates is equal.  What we trade is complexity.

**Comment by donmccurdy** [comment]
I see – thanks! No concerns about the exports in that case, changing them to reflect how they already get shipped in the bundle sounds good.

**Comment by donmccurdy** [comment]
I see – thanks! No concerns about the exports in that case, changing them to reflect how they already get shipped in the bundle sounds good.

**Comment by marcofugaro** [comment]
Yes this PR does not affect tree-shaking, nor does it make it better.

My point is that this PR changes the exports in a confusing way for no apparent reason, since the import/exports get compiled away. I don't see how this change is beneficial for future maintainers to understand what is going on. The real issue is with eslint, a tool shouldn't dictate how we structure our code.

But I guess this can be overlooked one since it's supposed to be temporary.

**Comment by donmccurdy** [comment]
We want to have automated warnings when our source code is not ES2018 compliant. I would've ideally excluded imports/exports from that, since they're handled by the build process anyway, but that path doesn't seem to be practical, so I'm fine with changing the imports/exports. We can clean them up later if needed.

It's not just about 1-2 unsupported ES2020 features — we want our code to be ES2018-compliant, and we don't want to be surprised by a bit of newer syntax sneaking in, or surprised by transpilers doing something we didn't expect. I worry that just enforcing ES2020 and adding a custom rules on top will not give us those guarantees, is that wrong?

**Comment by epreston** [comment]
I'm confused.  Changing the only 3 imports that are using a different syntax to the other 100 adds confusion ?  The lines that created confusion about tree-shaking ?

The project agrees on a syntax level, they is set in one line, PR's get validated against it, no gaps.  You set any tool to 2018 and you get the same results.  Seems kinda clean having it configured according to the manual in the most typical way.

**Comment by marcofugaro** [comment]
Hmm yeah I guess there is no way of having full ES2018 coverage without changing the import as well.

**Comment by mrdoob** [comment]
Ready to merge? 🧐

**Comment by marcofugaro** [comment]
Go ahead!

File: src/math/MathUtils.js
Comment (sunag) [inline src/math/MathUtils.js:300]: Would it not be better use just vars instead of `prop: var` like?

```js
const MathUtils = {
	DEG2RAD,
	RAD2DEG,
	generateUUID,
	...
```

Comment (epreston) [reply src/math/MathUtils.js:300]: It's exactly the same thing and gets translated to the code referenced.  This style allows the internal names to be changed (many modules import one or two directly bypassing this), while ensuring the documented API is always the same.

Comment (epreston) [reply src/math/MathUtils.js:300]: Last bit, when it comes to Terser and minification, these interface objects reduce the bundle size.  When the public API is used exclusively, the function names themselves can be condensed into a 2 character symbol, and used that way internally.  This marks the only spot where that 2 character replacement needs to appear in its full form.

Comment (sunag) [reply src/math/MathUtils.js:300]: I think it's cleaner than replicating the same names all times. For this specific case, I don't see the need to use it `prop:prop` instead of `prop`, or do we have some compatibility issue here?

Comment (epreston) [reply src/math/MathUtils.js:300]: Minification issue, sorry, I wrote that out just as you were responding.

So if any function names change, you have to introduce this object and it will be created anyway.  I'm just trying to avoid creating foot-guns.

Comment (LeviPesin) [reply src/math/MathUtils.js:300]: > The alternative is if any function names change, you have to introduce this format to maintain the API, it will be created anyway if something is inline-d. 

But why we would change a function name? And, if we would decide that a new name is better than the old one, we will also update the external name.

`{ prop1, prop2, ... }` format is also used in other places throught the library.

Comment (epreston) [reply src/math/MathUtils.js:300]: What's the goal of this feedback ?

> But why we would change a function name? 

API's and function names changing with 10 version support seems to be the norm.  So much so, there's a file dedicated to it with new updates every release, `Three.legacy.js`.  

> { prop1, prop2, ... } format is also used in other places through-out the library.

I didn't write that other code.  I hope the tradeoffs and reasons were well considered.  Plenty of places where things are not as frequently imported / exported where it does not matter because it's not API.

Comment (LeviPesin) [reply src/math/MathUtils.js:300]: > What's the goal of this feedback ?

Because the code style is such in files like this, for example, BufferGeometryUtils. I think we can discuss it in another issue if it should be changed -- I suggest that you use the `{ prop1, prop2, ...` format for now.

Comment (epreston) [reply src/math/MathUtils.js:300]: See the reasoning above.

**Comment by epreston** [comment]
@mrdoob can we wrap this up ?  I'd like to move on to the next.

File: src/math/MathUtils.js
Comment (LeviPesin) [reply src/math/MathUtils.js:300]: I still don't agree. But I think @mrdoob and @Mugen87 should decide.

**Review by Mugen87 — APPROVED** [review]

File: src/math/MathUtils.js
Comment (epreston) [reply src/math/MathUtils.js:300]: Yes, the final minimized size.  

The idea of removing minified builds is crazy.  If you are importing a library into your website from a CDN, you're not shipping un-minified dependencies in production.  Seems kinda odd for people to use a CDN to improve load times but, then lose that time because the build does not have a minified distribution.

Maybe there should be a document or readme about why things are done.

**Pull request merged**
